### ISSUE TITLE
Categoricals encoding

### ISSUE THREAD

**Issue by Zethson** [theislab/ehrapy#42]
We need to support various encodings of categoricals and eventually NLP to allow for ML algorithms to be applied.
This issue will solely deal with the encoding of categorials. There are primarily two kinds of categoricals:
1. categorical nominal
2. categorical ordinal

Both of them require different kinds of encodings. Nominal encodings need to be expanded into multiple columns because there is no dependency among them. The ML algorithms are not allowed to think that one value is in relation to another one. Ordinal categoricals can have dependencies and therefore values can be in relation to another. 

Helpful article: https://www.analyticsvidhya.com/blog/2020/08/types-of-categorical-data-encoding/

- [x] The encoding function must have a simple interface and provide the ability to apply different encodings to different columns. Furthermore, it must support the selection of columns either via indices or column names.
- [x] The applied encodings must be saved as a dictionary in _uns_ to preserve some kind of history 
- [x] one-hot encoding
- [ ] target encoding
- [ ] backward difference (https://contrib.scikit-learn.org/category_encoders/backward_difference.html)
- [ ] catboost (https://contrib.scikit-learn.org/category_encoders/catboost.html)
- [x] count encoding (https://contrib.scikit-learn.org/category_encoders/count.html)
- [ ] GLM (https://contrib.scikit-learn.org/category_encoders/glmm.html)
- [ ] Helmert (https://contrib.scikit-learn.org/category_encoders/helmert.html)
- [ ] James Stein (https://contrib.scikit-learn.org/category_encoders/jamesstein.html)
- [ ] Leave one out (https://contrib.scikit-learn.org/category_encoders/leaveoneout.html)
- [ ] M-estimate (https://contrib.scikit-learn.org/category_encoders/mestimate.html)
- [ ] Polynomial (https://contrib.scikit-learn.org/category_encoders/polynomial.html)
- [ ] sum coding (https://contrib.scikit-learn.org/category_encoders/sum.html)
- [ ] weight of evidence (https://contrib.scikit-learn.org/category_encoders/woe.html)
- [x] label encoding
- [ ] ordinal encoding with range (https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html)

**Comment by Imipenem** [comment]
- Binary encoding does not make sense since it's similar to one-hot-encoding beside the fact it stores the encoding in bitstrings (which is not what we want, since we only ant numbers in the encoded X)
- BaseN is not a good idea because depending on the base one chooses, its either similar to an existing encoding mode (N = 1 is one hot encoding), or its binary encoding (N = 2) or its a vanilly ordinal encoding.

**Comment by Zethson** [comment]
Let's not forget the progressbars. Need fixes & need to be added for all encodings. They should just disappear when finished.

**Comment by Zethson** [comment]
Out of scope for now. Not required and rarely used.
